CL/HIER: bcast 2step algorithm

vspetrov opened the pull request:
## What
Adds 2 lvl pipelined broadcast algorithm to CL/HIER

Perf improvement. Example on 32nodes,  ppn=56.

UCC master   (tl/ucp) | UCC new (cl/hier, shm+ucp) | UCC new (cl/hier, shm+ucp) +   BCAST_KN_RADIX=6
-- | -- | --
6.46 | 5.27 | 4.38
6.49 | 5.32 | 4.47
6.51 | 5.32 | 4.39
6.46 | 5.33 | 4.44
6.45 | 5.29 | 4.39
7.21 | 5.54 | 4.59
8.19 | 6.3 | 5.21
8.76 | 7.2 | 6.42
10.95 | 7.6 | 6.15
11.72 | 8.25 | 6.55
14.06 | 9.15 | 7.44
16.86 | 11.78 | 9.72
24.28 | 17.73 | 14.92

Sergei-Lebedev commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,267 @@
+                         ucc_base_coll_args_t *coll_args, ucc_base_team_t *team,
+                         ucc_coll_task_t **task)
+{
+    ucc_cl_hier_team_t *cl_team   = ucc_derived_of(team, ucc_cl_hier_team_t);
```
alignment

Sergei-Lebedev commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,267 @@
+    int                       n_frags, pipeline_depth;
+    ucc_status_t              status;
+
+    if (coll_args->args.op == UCC_OP_AVG) {
```
whe checking UCC_OP_AVG?

vspetrov replied:
bad copy-paste

Sergei-Lebedev commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,267 @@
+
+    if (ucc_unlikely(status != UCC_OK)) {
+        cl_error(team->context->lib,
+                 "failed to init pipelined 2step ar schedule");
```
ar -> bcast?

Sergei-Lebedev commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,267 @@
+
+    schedule->super.super.super.post = ucc_cl_hier_2step_bcast_start;
+    schedule->super.super.super.triggered_post = ucc_triggered_post;
+    schedule->super.super.super.finalize = ucc_cl_hier_ar_2step_schedule_finalize;
```
ar->bcast?

Sergei-Lebedev commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,267 @@
+    return status;
+}
+
+static ucc_status_t ucc_cl_hier_ar_2step_schedule_finalize(ucc_coll_task_t *task)
```
ar->bcast?

Sergei-Lebedev commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,267 @@
+    ucc_cl_hier_team_t *cl_team = ucc_derived_of(team, ucc_cl_hier_team_t);
+    ucc_team_t         *core_team = team->params.team;
+    ucc_coll_task_t    *tasks[2] = {NULL, NULL};
+    int root                        = coll_args->args.root;
```
ucc_rank_t

Sergei-Lebedev commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,267 @@
+    ucc_team_t         *core_team = team->params.team;
+    ucc_coll_task_t    *tasks[2] = {NULL, NULL};
+    int root                        = coll_args->args.root;
+    int rank                        = UCC_TL_TEAM_RANK(cl_team);
```
ucc_rank_t

Sergei-Lebedev commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,267 @@
+
+    if (SBGP_ENABLED(cl_team, NODE)) {
+        args.args.root = root_on_local_node
+            ? find_root_node_rank(root, cl_team) : 0;
```
is it guaranteed that node leader rank always has rank 0 in node group?

vspetrov replied:
good point. it is true now because parameter node_leader_rank_id is set to 0 during topo init. I changed the code to use that correct value.

Sergei-Lebedev commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,267 @@
+        ucc_derived_of(task, ucc_schedule_pipelined_t);
+
+    cl_debug(task->team->context->lib,
+             "posting 2step ar, buf %p, count %zd, dt %s"
```
ar->bcast?

Sergei-Lebedev commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,267 @@
+    int          n_frags = sp->super.n_tasks;
+    ucc_status_t status;
+
+    status = ucc_cl_hier_bcast_2step_init_schedule(coll_args, team, frag_p,
```
status is not really used, can do just
```
return ucc_cl_hier_bcast_2step_init_schedule(...)
```

Sergei-Lebedev commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,267 @@
+                                        ucc_schedule_t **sched_p, int n_frags)
+{
+    ucc_cl_hier_team_t *cl_team = ucc_derived_of(team, ucc_cl_hier_team_t);
+    ucc_team_t         *core_team = team->params.team;
```
alignment

Sergei-Lebedev commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,267 @@
+    int                  n_tasks, i, first_task;
+
+    args           = *coll_args;
+    n_tasks        = 0;
```
init vars at declaration

Sergei-Lebedev commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,267 @@
+#include "core/ucc_team.h"
+#include "../cl_hier_coll.h"
+
+#define MAX_AR_2STEP_TASKS 3
```
unused

Sergei-Lebedev commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,267 @@
+static inline ucc_rank_t
+find_root_node_rank(ucc_rank_t root, ucc_cl_hier_team_t *cl_team)
+{
+    ucc_rank_t  node_rank  = UCC_RANK_INVALID;
```
extra space before =

Sergei-Lebedev commented on src/components/cl/hier/bcast/bcast.h:
```diff
@@ -0,0 +1,39 @@
+enum
+{
+    UCC_CL_HIER_BCAST_ALG_2STEP,
+    UCC_CL_HIER_BCAST_ALG_SPLIT_RAIL,
```
what is bcast split rail?

vspetrov replied:
nothing

samnordmann commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,267 @@
+    }
+
+    if (SBGP_ENABLED(cl_team, NODE_LEADERS)) {
+        args.args.root = find_root_net_rank(ucc_team_rank_host_id(root, core_team), cl_team);
```
text width exceeds 80

samnordmann commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,267 @@
+        n_tasks++;
+    }
+
+    ucc_task_subscribe_dep(&schedule->super, tasks[first_task], UCC_EVENT_SCHEDULE_STARTED);
```
text width

samnordmann commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,267 @@
+    ucc_coll_task_t    *tasks[2] = {NULL, NULL};
+    int root                        = coll_args->args.root;
+    int rank                        = UCC_TL_TEAM_RANK(cl_team);
+    int root_on_local_node          = ucc_team_ranks_on_same_node(root, rank, core_team);
```
text width

samnordmann commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,267 @@
+            coll_args, team, (ucc_schedule_t **)task, n_frags);
+    }
+
+    schedule = ucc_cl_hier_get_schedule(cl_team);
```
initialize variable at the top of the function?

vspetrov commented on src/components/cl/hier/bcast/bcast.h:
```diff
@@ -0,0 +1,39 @@
+enum
+{
+    UCC_CL_HIER_BCAST_ALG_2STEP,
+    UCC_CL_HIER_BCAST_ALG_SPLIT_RAIL,
```
split rail is not implemented in this PR

Sergei-Lebedev commented:
@vspetrov looks like with new gtest asan found memleak, do you want to fix it in this PR. Or I can create fix in another PR

Sergei-Lebedev commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,264 @@
+
+    for (i = 0; i < sbgp->group_size; i++) {
+        if (ucc_ep_map_eval(sbgp->map, i) == root) {
+            node_rank = i;
```
just return i;

Sergei-Lebedev commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,264 @@
+            break;
+        }
+    }
+    return node_rank;
```
return UCC_RANK_INVALID, node_rank variable is not needed

Sergei-Lebedev commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,264 @@
+    for (i = 0; i < sbgp->group_size; i++) {
+        rank = ucc_ep_map_eval(sbgp->map, i);
+        if (ucc_team_rank_host_id(rank, core_team) == root_host_id) {
+            net_rank = i;
```
return i

Sergei-Lebedev commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,264 @@
+            break;
+        }
+    }
+    return net_rank;
```
return ucc_rank_invalid

Sergei-Lebedev commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,264 @@
+            goto out;
+        }
+        n_tasks++;
+        if (root_on_local_node && (root != rank)) {
```
seems like this condition is always false
if root != rank => root on remote node because we are inside if (SBGP_ENABLED(cl_team, NODE_LEADERS))
if root_on_local_node => root == rank because of the same reason

In that case you don't need first_task because first_task is always 0

vspetrov replied:
Yes, i beleive you are right. I re-checked and tested.

Sergei-Lebedev commented on src/components/cl/hier/bcast/bcast_2step.c:
```diff
@@ -0,0 +1,255 @@
+    ucc_schedule_add_task(schedule, tasks[0]);
+    if (n_tasks > 1) {
+        if (root == rank) {
+            ucc_task_subscribe_dep(&schedule->super, tasks[1],
```
it's not correct, you can start 2 tasks (node leaders bcast and node bcast) in parallel only if root rank belongs to node leaders group.

vspetrov replied:
> it's not correct, you can start 2 tasks (node leaders bcast and node bcast) in parallel only if root rank belongs to node leaders group.

If root is not node leader then it has only 1 task, and will not get there

Sergei-Lebedev replied:
ok, but node leader task should depend on node task in that case

vspetrov replied:
right.. this is fun. that is exactly why i had "first_task" before. and the condition above was correct actually to handle that case. i'll revert that last one change

vspetrov commented:
so, i added "first_task" back since it handles exactly the required case. Also, the reason gtest didn't catch the bug when i removed "first_task" is because the was incorrect "reset" callback in gtest bcast unit test. fixed that as well.

vspetrov commented:
looks like persistent gtest hangs with new bcast 2 step. need to debug...

manjugv commented:
Need to check with UCF folks on the copyright

manjugv commented:
@vspetrov Please add the copyright blurb you want. It was cleared by UCF.
